File: scripts/validate-dialects.py
#!/usr/bin/env python3
"""Render every runnable example for every dialect and syntax-check it with sqlglot.

This is intentionally parse-only. PostgreSQL and MySQL are executed with real
service containers in the CI workflow; Trino/Hive are validated here until
service-container execution is added for them.
"""
from __future__ import annotations

import subprocess
import sys
from pathlib import Path

import sqlglot

ROOT = Path(__file__).resolve().parents[1]
EXAMPLES = [
    "examples/adults.tetaue",
    "examples/case.tetaue",
    "examples/joins.tetaue",
    "examples/lpbirthday.tetaue",
    "examples/orders.tetaue",
    "examples/report.tetaue",
    "examples/selective.tetaue",
    "examples/strings.tetaue",
    "examples/lib-project/main.tetaue",
]
DIALECTS = {
    "sqlite": "sqlite",
    "postgresql": "postgres",
    "mysql": "mysql",
    "trino": "trino",
    "hive": "hive",
}

failures: list[str] = []
for rel in EXAMPLES:
    path = ROOT / rel
    for tetaue_dialect, sqlglot_dialect in DIALECTS.items():
        proc = subprocess.run(
            ["bun", "run", "src/cli.ts", "render", str(path), "--dialect", tetaue_dialect, "--format", "compact"],
            cwd=ROOT,
            text=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
        if proc.returncode != 0:
            failures.append(f"{rel} ({tetaue_dialect}): render failed:\n{proc.stderr.strip()}")
            continue
        sql = proc.stdout.strip()
        try:
            sqlglot.parse_one(sql, read=sqlglot_dialect)
        except Exception as exc:  # sqlglot raises a variety of parser errors
            failures.append(f"{rel} ({tetaue_dialect}): sqlglot parse failed:\n{sql}\n{exc}")

if failures:
    print("\n\n".join(failures))
    sys.exit(1)
print(f"Validated {len(EXAMPLES)} examples across {len(DIALECTS)} SQL dialects.")

File: scripts/validate-dialects.ts
/******************************************************************************
 * Render every runnable example for every dialect and syntax-check it with
 * node-sql-parser.
 *
 *   bun run scripts/validate-dialects.ts
 *
 * This is intentionally parse-only. PostgreSQL and MySQL are executed with
 * real service containers in the CI workflow; Trino/Hive are validated here
 * until service-container execution is added for them.
 ******************************************************************************/
import { execFileSync } from 'node:child_process';
import { resolve } from 'node:path';
import { Parser } from 'node-sql-parser';

const ROOT = resolve(new URL('.', import.meta.url).pathname, '..');
const CLI = 'src/cli.ts';
const EXAMPLES = [
    'examples/adults.tetaue',
    'examples/case.tetaue',
    'examples/joins.tetaue',
    'examples/lpbirthday.tetaue',
    'examples/orders.tetaue',
    'examples/report.tetaue',
    'examples/selective.tetaue',
    'examples/strings.tetaue',
    'examples/lib-project/main.tetaue',
];
const DIALECTS: { [tetaue: string]: string } = {
    sqlite: 'sqlite',
    postgresql: 'postgresql',
    mysql: 'mysql',
    trino: 'trino',
    hive: 'hive',
};

const parser = new Parser();
const failures: string[] = [];
for (const rel of EXAMPLES) {
    const path = resolve(ROOT, rel);
    for (const [tetaueDialect, parserDialect] of Object.entries(DIALECTS)) {
        let sql: string;
        try {
            sql = execFileSync(
                'bun',
                ['run', CLI, 'render', path, '--dialect', tetaueDialect, '--format', 'compact'],
                { cwd: ROOT, encoding: 'utf8' },
            ).trim();
        } catch (err) {
            const detail = err instanceof Error ? err.message.trim() : String(err);
            failures.push(`${rel} (${tetaueDialect}): render failed:\n${detail}`);
            continue;
        }
        try {
            parser.astify(sql, { database: parserDialect });
        } catch (err) {
            const detail = err instanceof Error ? err.message : String(err);
            failures.push(`${rel} (${tetaueDialect}): node-sql-parser parse failed:\n${sql}\n${detail}`);
        }
    }
}

if (failures.length > 0) {
    console.log(failures.join('\n\n'));
    process.exit(1);
}
console.log(`Validated ${EXAMPLES.length} examples across ${Object.keys(DIALECTS).length} SQL dialects.`);
